diff -up git-2.38.1/Makefile.1~ git-2.38.1/Makefile
--- git-2.38.1/Makefile.1~	2022-10-19 00:54:24.334645798 +0200
+++ git-2.38.1/Makefile	2022-10-19 00:54:47.817825346 +0200
@@ -570,8 +570,8 @@ perllibdir_relative = $(patsubst $(prefi
 export prefix bindir sharedir sysconfdir perllibdir localedir
 
 # Set our default programs
-CC = cc
-AR = ar
+CC = $(CC)
+AR = $(AR)
 RM = rm -f
 DIFF = diff
 TAR = tar
