diff -up git-2.50.0/gitk-git/gitk.omv~ git-2.50.0/gitk-git/gitk
--- git-2.50.0/gitk-git/gitk.omv~	2025-07-04 23:08:34.316305554 +0200
+++ git-2.50.0/gitk-git/gitk	2025-07-04 23:13:55.057174020 +0200
@@ -12453,9 +12453,9 @@ namespace import ::msgcat::mc
 ::msgcat::mcload $gitk_msgsdir
 
 # First check that Tcl/Tk is recent enough
-if {[catch {package require Tk 8.4} err]} {
+if {[catch {package require Tk 9.0} err]} {
     show_error {} . [mc "Sorry, gitk cannot run with this version of Tcl/Tk.\n\
-                         Gitk requires at least Tcl/Tk 8.4."]
+                         Gitk requires at least Tcl/Tk 9.0."]
     exit 1
 }

diff --git a/gitk-git/gitk b/gitk-git/gitk
index 1968976..03a6ad3 100755
--- a/gitk-git/gitk
+++ b/gitk-git/gitk
@@ -7802,7 +7802,7 @@ proc gettree {id} {
             set treepending $id
             set treefilelist($id) {}
             set treeidlist($id) {}
-            fconfigure $gtf -blocking 0 -encoding binary
+            fconfigure $gtf -blocking 0 -translation binary
             filerun $gtf [list gettreeline $gtf $id]
         }
     } else {
@@ -8065,7 +8065,7 @@ proc gettreediffs {ids} {

     set treepending $ids
     set treediff {}
-    fconfigure $gdtf -blocking 0 -encoding binary
+    fconfigure $gdtf -blocking 0 -translation binary
     filerun $gdtf [list gettreediffline $gdtf $ids]
 }

@@ -8185,7 +8185,7 @@ proc getblobdiffs {ids} {
         error_popup [mc "Error getting diffs: %s" $err]
         return
     }
-    fconfigure $bdf -blocking 0 -encoding binary -eofchar {}
+    fconfigure $bdf -blocking 0 -translation binary -eofchar {}
     set blobdifffd($ids) $bdf
     initblobdiffvars
     filerun $bdf [list getblobdiffline $bdf $diffids]

 
